### ISSUE TITLE
[$500] Web -  When we send link with multiple ending parenthesis the link ends in the middle

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#26788]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to any report
2. Send a link with multiple umatched closing parenthesis that are not all at the end, for example: google.com/(toto))titi)

## Expected Result:
The link should end after google.com/(toto) because the first unmatched parenthesis is probably not part of the link, for example it could be the parenthesis closing the markdown: [google.com/(toto)](http://google.com/(toto))

## Actual Result:
The link ends after the unmatched parenthesis at a weird place

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.63.0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/66e95a6d-6560-4323-aa4a-2582e6c330ff

https://github.com/Expensify/App/assets/93399543/f88195fc-3e0d-4eaa-9fd7-d59d5fbb807f

**Expensify/Expensify Issue URL:**
**Issue reported by:** @ShogunFire
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1693485959926909

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~010ba033b476889ccc</li>
        <li>Upwork Job ID: 1700291645262581760</li>
        <li>Last Price Increase: 2023-09-08</li>
<li>Automatic offers: </li>
<ul>
<li>ShogunFire | Contributor | 26746827</li>
<li>ShogunFire | Reporter | 26746828</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @robertKozik</details>

**Comment by kbecciv** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
The link look wrong when we send a link with multiple closing parenthesis.
### What is the root cause of that problem?
We remove extra closing parenthesis from links here: https://github.com/Expensify/expensify-common/blob/7735de14112a968fd6a4f5af710d2fbaefc8809d/lib/ExpensiMark.js#L456-L471
But the way we do it is wrong because we just count the number of extra parenthesis x and then remove x characters from the link without checking if the closing parenthesis are at the end or not.
### What changes do you think we should make in order to solve the problem?
We should remove every characters after an unmatched closing parenthesis
This is the code I propose:
```
// remove everything characters after an unmatched ) parenthese (including the parenthese)
let brace = 0;
for (let i = 0; i < match[2].length; i++) {
    if (match[2][i] === '(') {
        brace++;
    }
    if (match[2][i] === ')') {
        if(brace <= 0){ // unmatched closing parenthesis
            const numberOFCharsToRemove = match[2].length - i;
            match[0] = match[0].substr(0, match[0].length - numberOFCharsToRemove);
            match[2] = match[2].substr(0, match[2].length - numberOFCharsToRemove);
            break;
        }
        brace--;
    }
}
```
### What alternative solutions did you explore? (Optional)

**Comment by alexpensify** [comment]
When I was reviewing for dupes, I found this GH: https://github.com/Expensify/App/issues/25532#issuecomment-1702686658

@ShogunFire are you addressing this issue in that GH listed above? I see you included similar links. 

cc: @MonilBhavsar and @0xmiroslav

**Comment by ShogunFire** [comment]
@alexpensify I found this issue while resolving the other one but I believe this is a different issue. The other one is about removing special characters from the url that we are not currently removing.
This one is about a problem with how we remove the end parenthesis currently.

**Comment by alexpensify** [comment]
Thanks for the feedback @ShogunFire! 

@robertKozik could you please review the proposal and identify if it will resolve this issue? Thanks!

**Comment by ShogunFire** [comment]
I couldn't edit my proposal but another approach that I put in the slack thread is to remove the unmatched end parenthesis only if they are at the end but I don't recomment it.

**Comment by khanumar03** [comment]
hi is issue still open ?

**Comment by khanumar03** [comment]
@ShogunFire! so we  want to normalize parenthesis ?

**Comment by MonilBhavsar** [comment]
@alexpensify I prefer having two issues separate. We're discussing in the other issue.

**Comment by robertKozik** [comment]
Hi all! Thanks for proposal @ShogunFire, I've checked it and it solves the issue. If we are ok with reducing the link to the first unmatched parenthesis, proposal is good to go

🎀 👀 🎀 C+ reviewed

**Comment by alexpensify** [comment]
@cead22 - any update on the proposal review? Thanks!

**Comment by cead22** [comment]
Proposal looks good to me, here's a few small notes
- Let's explain _why_ we're doing this in the commend, eg, _we want to end the link at the last closing parenthesis that matches an opening parenthesis
- Use parenthesis (singular) and parentheses (plural)
- Use `unmatchedOpenParentheses` instead of `brace`
- Use `else if` for the second condition
- Maybe (NAB) add a variable with a good name for `match[2]`

**Comment by ShogunFire** [comment]
Thanks, here is the PR for expensify-common https://github.com/Expensify/expensify-common/pull/578

**Comment by alexpensify** [comment]
Looks like the PR is waiting for @cead22 to review

**Comment by cead22** [comment]
Review submitted. Requested a couple of small changes

**Comment by alexpensify** [comment]
@ShogunFire have you been able to review the changes that Carlos suggested?

**Comment by alexpensify** [comment]
Thank you for the update!

**Comment by alexpensify** [comment]
Easy melvin, there is a PR in review here. I've added the `Reviewing` label to clarify the state of this one.

**Comment by alexpensify** [comment]
Looks like the PR was merged so waiting for the automation here.

**Comment by alexpensify** [comment]
Looks like the PR was merged last week, I need to figure out if automation failed here and identify the new payment date. 🤔

**Comment by alexpensify** [comment]
@cead22 and @robertKozik -- can you share an update on the status of this GH? I'm trying to figure out if the automation broke here or if we are waiting on another PR. Thanks!

**Comment by cead22** [comment]
This is done, but the PR was on expensify-common, not App, which is probably why the automation failed. There should be an App PR to bump the version of expensify-common, but I bet we've already updated it to a version that includes the fix made in expensify-common for this

**Comment by alexpensify** [comment]
@ShogunFire and @robertKozik - was there an App PR here? I don't see it and trying to make sure we pay everyone since there has been no regression. Thanks!

**Comment by ShogunFire** [comment]
@alexpensify No, I haven't done an App PR, I am not sure what I am supposed to do. @cead22 seems to say it might already be done

**Comment by cead22** [comment]
Yes this is done. The expensify-common PR was merged in Sep and the expensify-common commit that App is pointing to is from Oct

**Comment by robertKozik** [comment]
The last bump of the `expensify-common` is from last week, where PR from this issue was merged about a month.

**Comment by alexpensify** [comment]
@robertKozik any update on what action is needed here?

**Comment by robertKozik** [comment]
As @cead22 [comment](https://github.com/Expensify/App/issues/26788#issuecomment-1762002312), the PR is already live on staging/production. The PR for `expensify-common` has been merged in Sep. We didn't create the PR for updating the `expenfisy-common` version, but It has been done by somebody else in the meantime (lates update is from the last week, so it for sure contains our fix as well). As there is no trace of any regression I think we can proceed with this issue like the normal one which didn't cause any problems

**Comment by cead22** [comment]
@alexpensify do you know what comes next or how to trigger the automation for the payments?

**Comment by alexpensify** [comment]
I'm OOO today. On Monday I will review the GH for the payment summary (ex. is there is an urgency bonus) and start the process in Upwork manually.

**Comment by alexpensify** [comment]
Here is the payment summary: 

-  External issue reporter
-  Contributor that fixed the issue
-  Contributor+ that helped on the issue and/or PR

Upwork Job: 


*If applicable, the bonuses will be applied on the final payment 

Extra Notes regarding payment:

**Comment by alexpensify** [comment]
@ShogunFire - I sent over two Upwork contracts and need your approval to continue the process. Thanks!

**Comment by alexpensify** [comment]
Closing - I've completed the payment process in Upwork.
